source/Calamari.Aws/Integration/S3/S3ObjectExtensions.cs

```diff
@@ -4,16 +4,98 @@
+                .WithHeaderFrom("Cache-Control", headers => headers.CacheControl)
+                .WithHeaderFrom("Content-Disposition", headers => headers.ContentDisposition)
+                .WithHeaderFrom("Content-Encoding", headers => headers.ContentEncoding)
+                .WithHeaderFrom("Expires")
+                .WithHeaderFrom("x-amz-website​-redirect-location")
+                .WithHeaderFrom("x-amz-object-lock-mode");
+
+        private static T WithHeaderFrom<T>(this T values, string key) where T : IDictionary<string, Func<HeadersCollection, string>>
+        {
+            values.Add(key, (headers) => headers[key]);
+            return values;
+        }
+
+        private static T WithHeaderFrom<T>(this T values, string key, Func<HeadersCollection, string> accessor) where T : IDictionary<string, Func<HeadersCollection, string>>
+        {
+            values.Add(key, accessor);
+            return values;
+        }
+
+        public static IEnumerable<string> GetKnownSpecialHeaderKeys()
+        {
+            return SupportedSpecialHeaders.Keys;
+        }
+
+        public static bool MetadataEq(IDictionary<string, string> next, IDictionary<string, string> current)
+        {
+            var allKeys = next.Keys.Union(current.Keys).Distinct();
+            var (filtered, excluded) = SelectWithExclusions(allKeys, (val) => next.ContainsKey(val) && current.ContainsKey(val));
+
+            return excluded.Count == 0 && filtered.All(x => string.Compare(next[x], current[x], StringComparison.CurrentCultureIgnoreCase) == 0);
+        }
+
+        public static IDictionary<string, string> ToDictionary(this MetadataCollection collection)
+        {
+            return collection.Keys.ToDictionary(key => key, key => collection[key]);
+        }
+
+        public static IDictionary<string, string> GetCombinedMetadata(HeadersCollection headers, MetadataCollection metadata)
+        {
+            var result = new Dictionary<string, string>();
+
+            foreach (var key in headers.Keys.Where(SupportedSpecialHeaders.ContainsKey))
+            {
+                result.Add(key, SupportedSpecialHeaders[key](headers));
+            }
+
+            foreach (var key in metadata.Keys)
+            {
+                result.Add(key, metadata[key]);
+            }
+
+            return result;
+        }
+
+        public static IDictionary<string, string> ToCombinedMetadata(this PutObjectRequest request)
+        {
+            return GetCombinedMetadata(request.Headers, request.Metadata);
+        }
+
+        public static IDictionary<string, string> ToCombinedMetadata(this GetObjectMetadataResponse response)
+        {
+            return GetCombinedMetadata(response.Headers, response.Metadata);
+        }
+
+        private static (IList<T> filtered, IList<T> excluded) SelectWithExclusions<T>(this IEnumerable<T> values, Func<T, bool> predicate)
```

Comment:
I'm admittedly feeling confused by this method. It's only used in a single place, and I wonder if you could inline and simplify it into a few simple linq expressions inside `MetadataEq`?

Reply:
I have moved the method inline for clarity and got rid of the aggregate call